src/efi-application.c

```diff
@@ -288,10 +290,20 @@ __tpm_event_efi_bsa_rehash(const tpm_event_t *ev, const tpm_parsed_event_t *pars
 		return tpm_event_get_digest(ev, ctx->algo->openssl_name);
 	}
 
+	/* The next boot can have a different kernel */
+	if (sdb_is_kernel(evspec->efi_application)) {
+		new_application = sdb_get_next_kernel();
+		if (new_application) {
+			memcpy(&evspec_clone, evspec, sizeof(evspec_clone));
+			evspec_clone.efi_application = strdup(new_application);
+			evspec = &evspec_clone;
+		}
+	}
+
 	if (ctx->use_pesign)
 		return __efi_application_rehash_pesign(ctx, evspec->efi_partition, evspec->efi_application);
 
-	return __efi_application_rehash_direct(parsed, ctx);
+	return __efi_application_rehash_direct(&evspec_clone, ctx);
```

Comment:
This looks like it will fail if we're hashing something other than a kernel. If sdb_is_kernel returns false, evspec_clone remains uninitialized. So better call __efi_application_rehash_direct(evspec, ctx)

Reply:
omg ... yes. Fixed